DM-7354 Initial alert producer and consumer using Kafka and Avro

mtpatter opened the pull request:
Producer ticket is DM-7453 and consumer ticket is DM-7454.  Attaching pull request to the relevant epic since they can't really be tested separately.

mtpatter committed:
add code files

mtpatter committed:
add docker files

ctslater commented on bin/sendAlertStream.py:
```diff
@@ -0,0 +1,50 @@
+    args = parser.parse_args()
+    topic = args.topic
+    ccdnum = args.ccd
+    alertnum = args.alertnum
```
Kind of unnecessary to break these out.

ctslater commented on bin/sendAlertStream.py:
```diff
@@ -0,0 +1,50 @@
+                        help='Name of Kafka topic stream to push to.')
+    parser.add_argument('ccd', metavar='ccdId', type=int,
+                        help='CCD ID number.')
+    parser.add_argument('alertnum', metavar='alertnum', type=int,
```
metavar is redundant in most of these usages.

ctslater commented on python/lsst/alert/stream/alertConsumer.py:
```diff
@@ -0,0 +1,78 @@
+
+        except KeyboardInterrupt:
+            sys.stderr.write('%% Aborted by user\n')
+            sys.exit
```
Needs to be `sys.exit()`. Repeated throughout.

ctslater commented on python/lsst/alert/stream/kafkaUtils.py:
```diff
@@ -0,0 +1,87 @@
+"""Utilities for talking to Kafka.
+"""
```
All of the functions in here are basically writing to the console, so I think this all belongs in the consumer.

ctslater commented on bin/printAlertStream.py:
```diff
@@ -0,0 +1,44 @@
+
+    # Start consumer and print alert stream
+    streamListener = alertConsumer.AlertConsumer(topic, **conf)
+    streamListener.printAlertStream(alert_schema)
```
I don't understand why the Consumer doesn't manage the schema for me. As-is, to consume the alerts I have to do a lot of setup work, and `AlertConsumer` just supplies me with a `while` loop.

mtpatter replied:
I had the consumer manage the schema at first, and then I took it out, because the consumer can listen to the stream without decoding messages and so doesn't need a schema.  That use case might only be for me though, so I can do benchmarking/timing without decoding.

ctslater reviewed (approved):
Two general issues:
- `alert_stream` is not a particularly good name, since it doesn't indicate at all what this particular piece of code does as part of the alert system. Something that conveys some specific functionality like `alert_testbed` might be more informative. It might be useful to think in terms of what else might go in `lsst.alert.*`?
- The current organization has a producer and a consumer which are both very thin, and basically shepherd data between various helper functions from a separate module.  I'd argue that grab-bag utility modules are not very easy to understand, and it would be nicer if a single class encapsulated both the translation to/from avro and then sending or receiving messages. E.g., all I should have to do is `AlertStreamConnector.send(json)` and `json_alerts = AlertStreamConnector.poll()` (where `json` I think is a dictionary that is json-compatible, but which could evolve to be something closer to raw DIASource/DIAObject records from Science Pipelines). As is, the code isn't very object-oriented.

I have made a few other code style comments; note that most of them apply to multiple files even though I only highlighted an individual usage.

timj commented:
Has the code been tested on Python 3?

mtpatter commented:
Yes, on Python 3.5.2.

mtpatter committed:
Reconfigure alert producer to encode each alert separately.

Previous alert producer encoded one template alert and sent to Kafka repeatedly.

mtpatter committed:
Rework consumer to put in Avro decoding.

Now AlertConsumer polls Kafka for new messages and can optionally decode with a given Avro schema.

mtpatter committed:
Consolidate utilities into consumer.

mtpatter committed:
Fix instructions and refresh Docker.

mtpatter commented:
Ok, thanks very much, @ctslater.  I will leave the name alert_stream for now - I think it is an appropriate name since this encompasses what you need to generate or listen to a stream of alerts and Kafka is specifically a stream processing platform.  

I've kept the producer and consumer separate still but pushed the Avro encoding/decoding closer to transmitting/receiving messages and scrapped the kafkaUtils.  Encoding/decoding can be turned off to transmit/receive raw (non-Avro) messages.  

The Avro encoding here is very slow- I will try and swap this out with something faster.  And so I've just left the avroUtils separate.

mtpatter committed:
Move decodeMessage into alertConsumer class.

mtpatter committed:
Pull decoding exception out of alertConsumer.

mtpatter committed:
Add exception for reaching end of partition.
